### PR TITLE
Adding PXE-based API discovery test

### PR DESCRIPTION
**Pull request by Griffin-Sullivan**
Creating new fixture for configuring discovery and rewriting an old test to use the new fixtures for discovery and provisioning. My plan is to keep converting tests in this file over to the new fixtures.

Depends on https://github.com/SatelliteQE/nailgun/pull/842

### CODE REVIEW

**Comment by Griffin-Sullivan** [comment]
trigger: test-robottelo
pytest: tests/foreman/api/test_discoveredhost.py::TestDiscoveredHost::test_positive_provision_pxe_host
nailgun: 842

**Comment by Griffin-Sullivan** [comment]
trigger: test-robottelo
pytest: tests/foreman/api/test_discoveredhost.py::TestDiscoveredHost::test_positive_provision_pxe_host
nailgun: 842

**Comment by Griffin-Sullivan** [comment]
```============================ test session starts =====================
platform linux -- Python 3.9.13, pytest-7.1.3, pluggy-0.13.1
shared_function enabled - OFF - scope:  - storage: file
rootdir: /home/gsulliva/Programming/robottelo, configfile: pyproject.toml
plugins: forked-1.3.0, services-2.2.1, cov-2.12.1, xdist-2.5.0, reportportal-5.1.2, ibutsu-2.2.4, mock-3.9.0
collected 3 items                                                                                                                                                                                                                                                     

tests/foreman/api/test_discoveredhost.py ...                                                                                                                                                                                                                    [100%]

========= 3 passed, 103 warnings in 3648.28s (1:00:48) =====================```

File: pytest_fixtures/component/discovery.py
Comment (jyejare) [inline pytest_fixtures/component/discovery.py:7]: This should go under `core/broker.py` than here as its returning the sat.

File: pytest_fixtures/component/discovery.py
Comment (jyejare) [inline pytest_fixtures/component/discovery.py:23]: You can use settings fixture that sets and unsets value for session/function(probably) scope.

File: pytest_fixtures/component/discovery.py
Comment (jyejare) [inline pytest_fixtures/component/discovery.py:40]: Looping and then again lookup by their name doesnt sounds like a good strategy, you can directly lookup by name of setting.

Check you can also use settings fixture here for multiple settings changes or we can update that fixture to consume multiple settings update.

File: pytest_fixtures/component/provision_pxe.py
Comment (jyejare) [inline pytest_fixtures/component/provision_pxe.py:217]: Do we really need 6GB? I think 2 Gb should be enough max 4 gb.

File: pytest_fixtures/component/provision_pxe.py
Comment (jyejare) [inline pytest_fixtures/component/provision_pxe.py:209]: Pease rename it to pxe_host or similar

File: tests/foreman/api/test_discoveredhost.py
Comment (jyejare) [inline tests/foreman/api/test_discoveredhost.py:78]:
```suggestion
def discovered_host_provisioned(channel, ksrepo):
```

**Review by jyejare — APPROVED** [review]
Great Improvement all together left few comments.

@Griffin-Sullivan since we are going for RHEV based discovery , I think the effort for libvirt_discovery.py helpers refactoring would be useless. See if you find anything usefull there or feel free to delete that module. (Outside the scope of this PR)

File: pytest_fixtures/component/provision_pxe.py
Comment (Griffin-Sullivan) [reply pytest_fixtures/component/provision_pxe.py:217]: I believe that was set by @ogajduse. Ondrej, do you think we should reduce this?

File: tests/foreman/api/test_discoveredhost.py
Comment (Griffin-Sullivan) [reply tests/foreman/api/test_discoveredhost.py:78]: Maybe `assert_discovered_host_provisioned`?

File: pytest_fixtures/component/discovery.py
Comment (Griffin-Sullivan) [reply pytest_fixtures/component/discovery.py:23]: Wouldn't that require me to parametrize every test with the settings I'm going to change?

File: tests/foreman/api/test_discoveredhost.py
Comment (ogajduse) [inline tests/foreman/api/test_discoveredhost.py:359]: I would rather keep the test stubbed.
@lhellebr, do we plan to test Discovery on Libvirt?

**Review by ogajduse — CHANGES_REQUESTED** [review]
Great to see manifester being used in provisioning tests!
I left a few comments where one of them comments on a design flaw that needs to be fixed before merging.

I would also like to point out the fact that `module_provisioning_sat` runs before the first test in every module so `satellite-installer` runs with every module. This bad design was done by me. That is why I recommend provisioning test owners to change the fixture scope to `session`.

I reviewed mostly the fixtures, not the test module. I would leave a deeper review up to @rplevka - the original author.

File: pytest_fixtures/component/provision_pxe.py
Comment (ogajduse) [reply pytest_fixtures/component/provision_pxe.py:217]: I would leave it as is. I experienced `No space left on a disk` when I ran it with 2GiB. It is better to have more space for ramdisk, especially in discovery where we do `kexec`.
Resources should not be a concern since automation was moved out from RHVM-02. We should have plenty of space there now.

File: pytest_fixtures/component/provision_pxe.py
Comment (Griffin-Sullivan) [reply pytest_fixtures/component/provision_pxe.py:207]: Whoops didn't mean to leave this change in. I can parametrize it.

File: pytest_fixtures/component/provision_pxe.py
Comment (Griffin-Sullivan) [reply pytest_fixtures/component/provision_pxe.py:224]: Checking in a host before provisioning has finished. In this test I end it when I see the host calling for the kickstart repo, since the rest of the coverage should be handled by other provisioning tests.

Comment (Griffin-Sullivan) [reply pytest_fixtures/component/provision_pxe.py:224]: A little more detail: So when you try to checkin the host, it will actually be removed in RHV, but you will get the error above. So we just pass the error.

File: pytest_fixtures/component/provision_pxe.py
Comment (Griffin-Sullivan) [reply pytest_fixtures/component/provision_pxe.py:227]: Is there a better way to do this?

File: tests/foreman/api/test_discoveredhost.py
Comment (Griffin-Sullivan) [reply tests/foreman/api/test_discoveredhost.py:359]: So my plan was to convert all of the Libvirt Discovery tests to the new test class I added `TestDiscoveredHost`. All of the helpers we used for Libvirt Discovery no longer work. See https://github.com/SatelliteQE/robottelo/pull/10187 for how I want to clean up the module.

File: pytest_fixtures/component/provision_pxe.py
Comment (ogajduse) [reply pytest_fixtures/component/provision_pxe.py:227]: I would forward your question to @JacobCallahan or eventually @jyejare who also reviewed this PR.

Why do you need to catch the exception anyway? Where does it come from? If it is an exception generated due to a failed AAP WF, can you implement such logic in the playbook?

File: pytest_fixtures/component/provision_pxe.py
Comment (ogajduse) [reply pytest_fixtures/component/provision_pxe.py:224]: Gotcha, the test logic and steps make sense to me. So if this error comes from AAP, can you catch the error there instead of here in robottelo?

File: tests/foreman/api/test_discoveredhost.py
Comment (ogajduse) [reply tests/foreman/api/test_discoveredhost.py:359]: I see that you plan to test Discovery on RHV. Does it make sense to test Discovery also on libvirt? I tagged @lhellebr as he spun one env-libvirt instance up a few months back if I am not mistaken.

File: pytest_fixtures/component/provision_pxe.py
Comment (Griffin-Sullivan) [reply pytest_fixtures/component/provision_pxe.py:224]: The error comes from subscription-manager unregister. It tries to reach out to the host, but it's unreachable. We attempted to use `ignore_errors` there, but that didn't change anything. The WF will continue to execute after that, but broker will still send you the "ignored" error.

File: tests/foreman/api/test_discoveredhost.py
Comment (Griffin-Sullivan) [reply tests/foreman/api/test_discoveredhost.py:359]: I don't think so, but I am interested to see what Lukas thinks.

File: pytest_fixtures/component/provision_pxe.py
Comment (ogajduse) [inline pytest_fixtures/component/provision_pxe.py:214]:
```suggestion
            host_class=ContentHost,
```

File: tests/foreman/api/test_discoveredhost.py
Comment (rplevka) [reply tests/foreman/api/test_discoveredhost.py:359]: I can't see how the change to underlaying virtualization adds any value to discovery tests.

File: pytest_fixtures/core/broker.py
Comment (rplevka) [reply pytest_fixtures/core/broker.py:246]: Technically - you are right. Economically - i would go with it.
If we happen to run into issues, then we could always separate them.

**Review by rplevka — APPROVED** [review]

File: tests/foreman/api/test_discoveredhost.py
Comment (Griffin-Sullivan) [reply tests/foreman/api/test_discoveredhost.py:74]: For this first test I just wanted to get what was necessary to assert that a host is discovered and provisioned. I think you are right though. Maybe I should leave this in, and I can revisit using them in this test and others.

**Comment by Griffin-Sullivan** [comment]
Everything is still working
```
==================== 6 passed, 108 warnings in 5433.49s (1:30:33) ====================
```

File: tests/foreman/api/test_discoveredhost.py
Comment (Griffin-Sullivan) [reply tests/foreman/api/test_discoveredhost.py]: Decided to add it back. Will implement in another PR.

**Review by ogajduse — APPROVED** [review]
Nice work, @Griffin-Sullivan!

**Pull request merged**
